content/posts/windows-containers.md

```diff
@@ -0,0 +1,231 @@
+
+* Let's create a pipeline by clicking ```"New Item"```, then enter a name for your pipeline job (i.e ```"win-lin-pipeline"```) and select ```"Pipeline"``` as your job type. 
+
+- Select: 
+
+    * Definition: ```"Pipeline script from SCM"```
+    * SCM: ```Git```
+    * Repository URL: ```https://github.com/mluyo3414org/pod-templates.git ```
+
+ [![](/img/windows-containers/pipeline-options.png)](/img/windows-containers/pipeline-options.png)
+
+- Click ```Save```
+
+Let's take a look at the repository structure in ```https://github.com/mluyo3414org/pod-templates```:
+
+```
+
+├── Jenkinsfile
+├── README.md
+├── linux
+│   └── nodejs-pod.yaml
+└── windows
+    └── dotnet-pod.yaml
+
+```
+- Both ```nodejs-pod.yaml``` and ```dotnet-pod.yaml``` are files describing the Kubernetes pod agents used in the Jenkinsfile. 
+
+- The ```dotnet-pod.yaml``` has two container definitions: a **Windows based** ```jnlp```(```jenkins/jnlp-agent:latest-windows```) and a ```windows-dotnet``` container (```mcr.microsoft.com/dotnet/core/sdk:2.1```). We need to overwrite the ```jnlp``` container in this pod since otherwise it will use the **Linux based** ```jnlp``` container defined under ```Jenkins --> Configuration --> Pod templates```. In this case, Jenkins was automatically configured with the Linux pod: ```jenkins/jnlp-slave:3.27-1```.
+
+```
+kind: Pod
+metadata:
+  name: windows
+spec:
+  containers:
+  - name: jnlp
+    image: jenkins/jnlp-agent:latest-windows
+    tty: true
+  - name: windows-dotnet
+    image: mcr.microsoft.com/dotnet/core/sdk:2.1
+    tty: true
+  nodeSelector:
+    agentpool: winage
+```
+
+- The ```nodejs-pod.yaml``` has the ```node```(nodeJS) container definition and will use the default Linux based ```jnlp``` (```jenkins/jnlp-slave:3.27-1```) mentioned before.
+
+```
+kind: Pod
+metadata:
+  name: nodejs-app
+spec:
+  containers:
+  - name: nodejs
+    image: node:slim
+    command:
+    - cat
+    tty: true
+  nodeSelector:
+    agentpool: linage
+```
+
+- Notice both pod yaml definitions use  ```nodeSelector``` ([nodeSelector](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/)) to decide where these pods should be scheduled. If this is not specified, the Kubernetes scheduler will provision the pods following the [default behavior](https://kubernetes.io/blog/2017/03/advanced-scheduling-in-kubernetes/) which could possibly start a pod in the a node with the wrong OS. The tags used for the nodeSelectors are the default tags assigned by Azure when specifyng the nodepool name: ``` agentpool : nameOfNodePool ```. To find the node tags you can use the command: ``` kubectl get nodes --show-labels ```. Other options to prevent scheduling errors are ```taints```  ([more info] (https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/)).
```

Comment:
Couple of typos here:
- unnecessary "the" in "which could possibly start a pod in *the* a node with the wrong OS"
- misspelling: "when *specifyng* the nodepool name"